cartographer/mapping/internal/3d/local_trajectory_builder_3d.cc

```diff
@@ -352,19 +352,25 @@ LocalTrajectoryBuilder3D::InsertIntoSubmap(
   if (motion_filter_.IsSimilar(time, pose_estimate)) {
     return nullptr;
   }
-  const auto local_from_gravity_aligned =
-      pose_estimate.rotation() * gravity_alignment.inverse();
-
-  std::vector<std::shared_ptr<const mapping::Submap3D>> insertion_submaps =
-      active_submaps_.InsertRangeData(filtered_range_data_in_local,
-                                      local_from_gravity_aligned);
-
-  const Eigen::VectorXf rotational_scan_matcher_histogram =
+  const Eigen::VectorXf rotational_scan_matcher_histogram_in_gravity =
       scan_matching::RotationalScanMatcher::ComputeHistogram(
           sensor::TransformPointCloud(
               filtered_range_data_in_tracking.returns,
               transform::Rigid3f::Rotation(gravity_alignment.cast<float>())),
           options_.rotational_histogram_size());
+
+  const auto local_from_gravity_aligned =
+      pose_estimate.rotation() * gravity_alignment.inverse();
+  const float local_yaw_from_gravity =
+      transform::GetYaw(local_from_gravity_aligned);
```

Comment:
This is potentially numerically unstable. A better approach would be to forward the histogram unchanged along with the transform to local frame, combine all transforms in the end (this is then from an approximately gravity aligned frame to a different approximately gravity aligned frame, so mostly a rotation in yaw with small roll/pitch component, thus numerically stable) and then rotate the histogram by the yaw from this transform.

Reply:
Good catch! I fixed it.